File: samples/Neutrino/src/Fragments.swift
import UIKit
import RenderNeutrino

extension UI.Fragments {
  /// A component can be formed by a simple function returning a simple node hierarchy.
  static func badge() -> UINodeProtocol {
    return UINode<UIView>(reuseIdentifier: "Badge") { config in
      config.set(\UIView.yoga.width, 32)
      config.set(\UIView.yoga.height, 32)
      config.set(\UIView.backgroundColor, Color.red)
      config.set(\UIView.yoga.margin, 4)
      config.set(\UIView.layer.cornerRadius, 16)
      config.set(\UIView.clipsToBounds, true)
    }
  }

  static func bodyLabel(text: String) -> UINodeProtocol {
    return UINode<UILabel>() { config in
      config.set(\UILabel.numberOfLines, 0)
      config.set(\UILabel.textColor, .white)
      config.set(\UILabel.text, text)
      config.set(\UILabel.font, UIFont.systemFont(ofSize: 13, weight: UIFont.Weight.light))
    }
  }
}

File: samples/Neutrino/src/Fragments.js
// private subfragments.

_CounterButton = (props) => {
  return UINode(UIButton, "button", {
      text: props.count.toString(),
      backgroundColorImage: palette.pink,
      padding: 8,
      cornerRadius: 12,
      textColor: palette.white,
      font: typography.mediumBold,
  }, null)
}

_CounterReloadLabel = () => {
  return UINode(UILabel, null, {
      text: "⌘ + R to reload the javascript fragment",
      margin: 8,
      textColor: palette.text,
      font: typography.smallBold,
  }, null)
}

_CounterBadge = (text) => {
  return UINode(UILabel, null, {
      backgroundColor: palette.white,
      width: 32,
      height: 32,
      cornerRadius: 16,
      margin: 4,
      text: text,
      textColor: palette.text,
      textAlignment: ui.textAlignment.center,
      font: typography.extraSmallBold,      
  }, null)
}

// the exported fragment.

ui.fragment.Counter = (props, size) => {
  const badges = []
  for (let i = 0; i < props.count; i++) {
    badges.push(_CounterBadge((i+1).toString()))
  }

  const badgesContainer = UINode(UIView, null, {
      flexDirection: row,
      flexWrap: wrap,
      marginTop: 10,
  }, badges)

  return UINode(UIView, null, {
      padding: 25,
      backgroundColor: palette.green,
      flexDirection: column,
      width: size.width,
  }, [
      _CounterReloadLabel(),
      _CounterButton(props),
      badgesContainer,
  ])
}

ui.fragment.TableHeader = (props, size) => {
  return UINode(UILabel, null, {
      text: 'Welcome to Render-Neutrino',
      font: typography.smallBold,
      textAlignment: ui.textAlignment.center,
      width: size.width,
      padding: 8,
      backgroundColor: palette.text,
      textColor: palette.white,
  }, null)
}
